### PR TITLE
Feature/ran

### PR DESCRIPTION
**Pull request by mzambrano1**

### CODE REVIEW

**Review by pbadillatorrealba — COMMENTED** [review]
I left a couple of comments, mainly documentation.
The most important issue in the code is making pytorch dependency.
I believe that we should not do it so that the package does not grow too much. 
There I left an option to do this with try catch, but you have to test it in a clean environment.

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+import numpy as np
+from wefe.utils import check_is_fitted
+from wefe.word_embedding_model import WordEmbeddingModel
+import torch.nn as nn
+import torch
+from copy import deepcopy
+
+
+class RepulsionAttractionNeutralization(BaseDebias):
+    """Repulsion attraction Neutralization method.
```
Comment (pbadillatorrealba) [inline wefe/debias/repulsion_attraction_neutralization.py:17]:
```suggestion
    """Repulsion Attraction Neutralization method.
```

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+    Fn(wd) = |cos(wd,g)|
+    Neutralization minimize the bias to a particular gender. With g being de bias direction.
+
+    The optimization is performed by formulating a single objective:
+    F(wd) =  λ1Fr(wd) + λ2Fa(wd) + λ3Fn(wd)
+
+
+        Examples
+        --------
+        The following example shows how to execute Double Hard Debias method that reduces bias in a word embedding model:
```
Comment (pbadillatorrealba) [inline wefe/debias/repulsion_attraction_neutralization.py:54]:
```suggestion
        The following example shows how to execute Repulsion Attraction Neutralization method that reduces bias in a word embedding model:
```

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    Fa(wd) = |cos(wd,w)-1|/2
+    Attraction atracts wd to the original vector w to minimize the loss of semantic
+
+    Fn(wd) = |cos(wd,g)|
+    Neutralization minimize the bias to a particular gender. With g being de bias direction.
+
+    The optimization is performed by formulating a single objective:
+    F(wd) =  λ1Fr(wd) + λ2Fa(wd) + λ3Fn(wd)
+
+
```
Comment (pbadillatorrealba) [inline wefe/debias/repulsion_attraction_neutralization.py:51]: Here I would also add the equivalence between the paper's definitions of Preserve and Debias set and WEFE's target and ignore.

**Review by pbadillatorrealba — COMMENTED** [review]
More comments.

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+            w_b,
+            w,
+            repulsion_set,
+            bias_direction,
+            self._objective_function,
+            weights,
+        )
+        optimizer = torch.optim.Adam(
+            ran.parameters(), lr=learning_rate
+        )  ## dejar el optimizador como parametro?
```
[the remark this answers is not in the document]
Comment (mzambrano1) [reply wefe/debias/repulsion_attraction_neutralization.py]: Just Adam used

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+            new_model_name = f"{model.name}_{self.criterion_name_}_debiased"
+        model.name = new_model_name
+
+        if self.verbose:
+            print("Done!")
+
+        return model
+
+
+class RAN(nn.Module):
```
[the remark this answers is not in the document]
Comment (mzambrano1) [reply wefe/debias/repulsion_attraction_neutralization.py]: done!

File: wefe/debias/repulsion_attraction_neutralization.py (hunk 3 above)
Comment (mzambrano1) [reply wefe/debias/repulsion_attraction_neutralization.py:123]: both sets included in the documentation

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+        return similar_words
+
+    def _get_repulsion_set(
+        self,
+        model: WordEmbeddingModel,
+        word: str,
+        bias_direction: np.ndarray,
+        theta: float,
+        n_neighbours: int,
+    ) -> List[np.ndarray]:
```
[the remark this answers is not in the document]
Comment (mzambrano1) [reply wefe/debias/repulsion_attraction_neutralization.py:260]: documentation added

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+        bias_direction: np.ndarray,
+        repulsion_set: torch.Tensor,
+        weights: List[float],
+    ) -> torch.Tensor:
+        w1, w2, w3 = weights
+        return (
+            self._repulsion(w_b, repulsion_set) * w1
+            + self._attraction(w_b, w) * w2
+            + self._neutralization(w_b, bias_direction) * w3
+        )  # W3????**
```
[the remark this answers is not in the document]
Comment (mzambrano1) [reply wefe/debias/repulsion_attraction_neutralization.py]: forgot to delete, now its done

File: wefe/debias/repulsion_attraction_neutralization.py [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+from copy import deepcopy
+
+
+class RepulsionAttractionNeutralization(BaseDebias):
+    """Repulsion attraction Neutralization method.
+
+    This method allow reducing the bias of an embedding model creating a transformation
+    such that the stereotypical gender information are minimized with minimal semantic offset
+
+    This method is binary because it only allows 2 classes of the same bias criterion,
```
[the remark this answers is not in the document]
Comment (mzambrano1) [reply wefe/debias/repulsion_attraction_neutralization.py:94]: more detailed added to the description

**Comment by pbadillatorrealba** [comment]
I have 2 more comments:

- Wouldn't it be better to define the target function in the RAN class, as another method or even part of the forward?
- The fit method is not compatible with the superclass. It is missing the debias_criterion_name parameter. It is probably also missing its functionality (give a name to the new embedding).

**Pull request merged**